File: src/filter.js
const QUALITY_ORDER = ['CAM', null, '480p', '720p', '1080p', '4K']
const QUALITY_RANK = new Map()
QUALITY_ORDER.forEach((q, i) => QUALITY_RANK.set(q, i))

function rankOf(quality) {
  return QUALITY_RANK.has(quality) ? QUALITY_RANK.get(quality) : 1 // treat unknown like null
}

function buildBlacklistRegex(keywords) {
  if (!keywords || !keywords.length) return null
  const escaped = keywords
    .map(k => k.trim())
    .filter(Boolean)
    .map(k => k.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'))
  if (!escaped.length) return null
  return new RegExp(`\\b(?:${escaped.join('|')})\\b`, 'i')
}

export function filterTorrents(torrents, config) {
  const maxQ = config.maxQuality
  const minQ = config.minQuality
  const maxBytes = config.maxSizeGb ? config.maxSizeGb * 1024 ** 3 : null
  const minBytes = config.minSizeMb ? config.minSizeMb * 1024 ** 2 : null
  const blRe = buildBlacklistRegex(config.blacklistKeywords)

  return torrents.filter(t => {
    const q = t.parsedTitle ? t.parsedTitle.quality : null
    if (maxQ && rankOf(q) > rankOf(maxQ)) return false
    if (minQ && rankOf(q) < rankOf(minQ)) return false
    if (maxBytes && Number.isFinite(t.size) && t.size > maxBytes) return false
    if (minBytes && Number.isFinite(t.size) && t.size < minBytes) return false
    if (blRe && blRe.test(t.title || '')) return false
    return true
  })
}

File: internal/filter/filter.go
package filter

import (
	"regexp"
	"strings"

	"github.com/lborruto/jackstream/internal/config"
	"github.com/lborruto/jackstream/internal/parser"
)

type Filterable interface {
	Title() string
	ParsedTitle() parser.Parsed
	Size() int64
}

// Quality rank: CAM=0, unknown=1, 480p=2, 720p=3, 1080p=4, 4K=5.
var qualityRank = map[string]int{
	"CAM": 0, "": 1, "480p": 2, "720p": 3, "1080p": 4, "4K": 5,
}

func rankOf(q string) int {
	if v, ok := qualityRank[q]; ok {
		return v
	}
	return 1 // unknown label treated as null
}

func buildBlacklistRegex(kw []string) *regexp.Regexp {
	if len(kw) == 0 {
		return nil
	}
	parts := make([]string, 0, len(kw))
	for _, k := range kw {
		k = strings.TrimSpace(k)
		if k == "" {
			continue
		}
		parts = append(parts, regexp.QuoteMeta(k))
	}
	if len(parts) == 0 {
		return nil
	}
	return regexp.MustCompile(`(?i)\b(?:` + strings.Join(parts, "|") + `)\b`)
}

func Filter(items []Filterable, c config.Config) []Filterable {
	maxQ, minQ := c.MaxQuality, c.MinQuality
	var maxBytes, minBytes int64
	if c.MaxSizeGb > 0 {
		maxBytes = int64(c.MaxSizeGb * float64(1<<30))
	}
	if c.MinSizeMb > 0 {
		minBytes = int64(c.MinSizeMb * float64(1<<20))
	}
	blRe := buildBlacklistRegex(c.BlacklistKeywords)

	out := make([]Filterable, 0, len(items))
	for _, it := range items {
		q := it.ParsedTitle().Quality
		if maxQ != "" && rankOf(q) > rankOf(maxQ) {
			continue
		}
		if minQ != "" && rankOf(q) < rankOf(minQ) {
			continue
		}
		size := it.Size()
		if maxBytes > 0 && size > maxBytes {
			continue
		}
		if minBytes > 0 && size < minBytes {
			continue
		}
		if blRe != nil && blRe.MatchString(it.Title()) {
			continue
		}
		out = append(out, it)
	}
	return out
}
